[CCA][Memory Snapshot] Move user_defined annotations to Native Caching Allocator

aaronenyeshi opened the pull request:
Summary: Instead of embedding the user_defined TraceEntry inside of device_traces, which causes issues when some threads may not have the proper device id set, save them into an external_annotations field by using a RingBuffer<AnnotationEntry> called annotation_buffer owned by the NativeCachingAllocator.

Test Plan: CI, resnet run, and FBR model.

Differential Revision: D59703213

Pulled By: aaronenyeshi

zdevito commented on torch/csrc/profiler/combined_traceback.h:
```diff
@@ -58,15 +58,19 @@ struct TORCH_API CapturedTraceback : public c10::GatheredContext {
 
-  void recordUserDefinedFrame(const unwind::Frame& frame) {
-    user_defined_frames_.push_back(frame);
+  void recordUserDefinedMetadata(std::string name, std::string value) {
```
It doesn't really make sense to put metadata here. This is suppose to be a unified traceback object. Adding extra frames to it made sense, but arbitrary extra data will be hard to treat consistently.

aaronenyeshi replied:
Makes sense, I've removed user_defined frames from traceback object now.

zdevito commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -105,16 +105,16 @@ void _initRecordAnnotations() {
-              unwind::Frame frame{fn.name(), "START", 0};
               auto r = std::make_shared<CapturedTraceback>();
-              r->recordUserDefinedFrame(frame);
+              r->recordUserDefinedMetadata("name", fn.name());
```
I think it would be possible to simplify here by making the user-defined metadata a different class compared to trace entry. In particular, it could just have a time stamp, and then a str -> str dictionary, which would allow all this information to go in one entry.

aaronenyeshi replied:
This is a great point, I've changed to a different struct for AnnotationEntry, which will only hold what it needs.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -3028,7 +3024,10 @@ class NativeCachingAllocator : public CUDAAllocator {
+  // Variables by memory snapshot
   c10::ApproximateClockToUnixTimeConverter clock_converter;
+  bool record_history = false;
+  TraceEntryRingBuffer annotation_buffer;
```
From comments below, it might make sense to have annotations be a different class that trace entries because they share nearly no fields and require extra fields. It should be pretty simple to reproduce or template out the ring buffer nature of this class to allow for that.

aaronenyeshi replied:
Sounds good, I've templated out the ring buffer, so we can re-use for a new struct AnnotationEntry.

zdevito reviewed (commented):
I think this resolves the issues with user annotations getting duplicated. Some inline suggestions to avoid having to change combined_traceback and enable more data to get into user annotations.

aaronenyeshi commented:
Fixing lintrunner one more time, all other signals are passing!
![image](https://github.com/user-attachments/assets/1f0b04dc-e96d-4357-baac-df3087d1f0a0)

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -798,47 +798,48 @@ cudaError_t cudaMallocMaybeCapturing(void** p, size_t size) {
+class CCARingBuffer {
  public:
-  TraceEntryRingBuffer() {
+  CCARingBuffer() {
```
nit: can just be RingBuffer, it is local to this file.

aaronenyeshi replied:
Good point! Changed

zdevito commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -395,6 +402,7 @@ std::string _memory_snapshot_pickled() {
   result.insert("segments", segments);
   result.insert("device_traces", traces);
   result.insert("allocator_settings", allocator_settings);
+  result.insert("user_defined", user_defined);
```
nit: maybe `external_annotations`, otherwise I ask "used-defined what?"

aaronenyeshi replied:
That makes sense! Changed

zdevito reviewed (approved):
Looks good!

aaronenyeshi commented:
Addressed a couple nits, CCARingBuffer renamed to RingBuffer and replaced user_defined with external_annotations.

aaronenyeshi committed:
[CCA][Memory Snapshot] Move user_defined annotations to Native Caching Allocator (#130964)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130964

Instead of embedding the user_defined TraceEntry inside of device_traces, which causes issues are some threads may not have the proper device id set, save them into a user_defined annotation_buffer owned by the NativeCachingAllocator.

Test Plan: CI, resnet run, and FBR model.

Differential Revision: D59703213

Pulled By: aaronenyeshi

aaronenyeshi commented:
@pytorchbot merge -i "passed all the signals before the nit changes (non-functional changes)"

aaronenyeshi commented:
@pytorchbot merge -i
